design-docs/improved-console.md

```diff
@@ -0,0 +1,283 @@
+Each`RegionType` is associated with a color. This allows us to show failures in arbitrary locations
+as needed with `--continue`
+
+### Test Coverage
+* Logs should be streamed with plain output and not throttled when not attached to a Console
+* The TAPI is unaffected except for additional BuildOperations now published
+* Logs should be streamed with plain output when attached Terminal lacks of color or cursor support
+* Renders well on default macOS Terminal, default Windows Console, Cygwin, PowerShell and common Linux Terminals
+
+### Open Issues
+* Characters outside ASCII will give a more polished look, but with risk of poor display if console font doesn't
+support them
+* ASCII-based, unstyled alternatives:
+```
+[##                        ] 6% Building
+[‡‡‡‡                      ] 12% Building
+‹===================       › 60% Building
+«==========================» 100% BUILD SUCCESS
+```
+
+## Story: Display intra-operation progress
+When we know in advance the number of things to be processed, we display progress in [Complete / All] format. 
+Provide APIs to optionally add # complete and # of all items.  
+
+For example:
+```
+[23 / 65] Configuring tasks
+[999 / 1234] Running tests
+[384 / 218932] Compiling classes
+[56 / 245] Resolving dependencies
+```
+
+### Implementation
+ TODO
+
+## Story: Customizable ProgressBarFormat
+Allow users to configure how `ProgressBar` is formatted. This allows for great creativity
+without needing to provide a default implementation
+
+Details that can be customized:
+ * `ProgressBarRegion` prefix
+ * `ProgressBarRegion` suffix
+ * `ProgressBar` complete char (default is "#")
+ * `ProgressBar` width (default is 50)
+
+### Implementation
+`ProgressBarFormatter` implementation would read the following properties and provide defaults.
+```bash
+org.gradle.console.progressbar.succeeded.prefix="[32m"   #FG_GREEN
+org.gradle.console.progressbar.failed.prefix="[31m"      #FG_RED
+org.gradle.console.progressbar.inprogress.prefix="[33m"  #FG_YELLOW
+org.gradle.console.progressbar.unstarted.prefix="[30m"   #FG_BLACK
+org.gradle.console.progressbar.succeeded.suffix=""
+org.gradle.console.progressbar.failed.suffix=""
+org.gradle.console.progressbar.inprogress.suffix=""
+org.gradle.console.progressbar.unstarted.suffix=""
+org.gradle.console.progressbar.completechar=" "
+org.gradle.console.progressbar.completechar="#"
+org.gradle.console.progressbar.width=50
+```
+
+### Test Cases
+* Helpful error message is displayed if progressbar.width isn't a natural number
+* `progressbar.completechar` accepts multi-byte characters and displays correctly
+* Helpful error message if `progressbar.completechar` isn't one character in length
+* Progress bar is truncated at given width
+* Fuzz testing to protect against any security issues arising from using user input
+
+### Open issues
+* Could this concept allow customization of ProgressLine format, providing even greater flexibility?
+
+## Story: Gradle-provided slick themes
```

Comment:
I'll respond to this in the design doc and with a demo, but there will be an ASCII based existing theme. However, this is a blocker for release IMO.